h/services/group.py

```diff
@@ -106,14 +105,22 @@ def groupids_readable_by(self, user, group_ids=None):
             )
             # Union these with the world readable ones
             # We wrap a subquery around in case we need to apply the group_ids filter below
-            query = select(
-                union(readable_by_world_query, user_is_member_query).subquery()
-            )
+            union_groups = union(
+                readable_by_world_query, user_is_member_query
+            ).subquery()
+            # We explicitly alias the union to hint SQLAlchemy to fetch Group objects
+            group_alias = aliased(Group, union_groups)
+            query = select(group_alias)
 
         if group_ids:
             query = query.where(literal_column("pubid").in_(group_ids))
 
-        return self.session.scalars(query).all()
+        return self.session.scalars(query)
+
+    def groupids_readable_by(self, user, group_ids=None):
+        """Return a list of groupids for which the user has read access."""
+
+        return [g.pubid for g in self.groups_readable_by(user, group_ids)]
```

Comment:
This query now loads entire groups rather than just their IDs from the DB. Could this have a performance impact for any existing queries that might cause a problem?

Reply:
I don't think this is going to make a big difference. `groupids_readable_by()` is [only called in one place](https://github.com/search?q=repo%3Ahypothesis%2Fh%20groupids_readable_by&type=code) which is the search API's `GroupFilter`:

https://github.com/hypothesis/h/blob/61e70bb51f45fad21bff73322937f82ea91157bf/h/search/query.py#L210-L226

`GroupFilter` is only used by the search API's `group` parameter. I think the `group` parameter can be used to specify multiple groups to filter by providing the query param multiple times, but it's mostly only used once.